### PR TITLE
DM-24254: Eliminate slow calexp retrievals.

### PR DESCRIPTION
**Pull request by ktlim**
In almost all cases, dataset components can be used to avoid reading pixels.  In one case, a single pixel is read instead in order to get full metadata, plus a slightly hacky method is used to load the mask from an exposure.

### CODE REVIEW

**Comment by laurenam** [comment]
Thanks for the fixes K-T!  It looks like another instance can be updated here: https://github.com/lsst-dm/pipe_analysis/blob/master/python/lsst/pipe/analysis/coaddAnalysis.py#L585

Also, do you happen to know of a better fallback to get the filter (if it was not included in the `dataId`): https://github.com/lsst-dm/pipe_analysis/blob/master/python/lsst/pipe/analysis/utils.py#L1243-L1244

**Comment by RobertLuptonTheGood** [comment]
Please do not assume you can get the filter from the dataId.  Treat them as opaque

**Comment by laurenam** [comment]
I don’t assume it, I `try` it and “fallback” to getting it via the `calexp` if it’s not there.

**Comment by ktlim** [comment]
The better way to get the filter is `calexp_filter` (or the equivalent for coadds).  I've made this change.  I'll look at the other analysis tasks next.

**Comment by laurenam** [comment]
Great.  I did do a search for other instances...but another set of eyes is always a good thing 😉

File: python/lsst/pipe/analysis/visitAnalysis.py [hunk 1]
```diff
@@ -550,22 +550,35 @@ def readCatalogs(self, dataRefList, dataset, repoInfo, aliasDictList=None, fakeC
...
-                ccdHeight = exp.getHeight()
+                ccdWidth = mask.getWidth()
+                ccdHeight = mask.getHeight()
 
                 ccdCorners = wcs.pixelToSky([lsstGeom.Point2D(0, 0), lsstGeom.Point2D(ccdWidth, ccdHeight)])
                 areaDict["corners_" + str(dataRef.dataId["ccd"])] = ccdCorners
 
                 # Check which fake sources fall on the ccd
+                # TODO: this looks a bit iffy, as a fake could still be on the
+                # CCD but outside where sources were detected
```
Comment (sr525) [inline python/lsst/pipe/analysis/visitAnalysis.py:581]: I'm going to be doing quite a bit of fakes plots/pipe_analysis faffing in the near future. If this is a problem it should show up then and I'll fix it.

File: python/lsst/pipe/analysis/visitAnalysis.py [hunk 2]
```diff
@@ -550,22 +550,35 @@ def readCatalogs(self, dataRefList, dataset, repoInfo, aliasDictList=None, fakeC
...
-                mask = exp.mask.array
-                maskBad = mask & 2**0
-                good = len(np.where((maskBad == 0))[0])
-                pixScale = exp.getWcs().getPixelScale().asArcseconds()
+
+                # Hacky but fast way of getting just the mask
+                # Actual mask is needed because BAD pixels are more than
+                # just the defects
+                fname = butler.get("calexp_filename", dataId)[0]
+                mask = afwImage.Mask(fname, hdu=2)
```
Comment (timj) [inline python/lsst/pipe/analysis/visitAnalysis.py:565]: If we didn't want to specify an hdu explicitly would `ExposureFitsReader.readMask` work better?

Comment (ktlim) [reply python/lsst/pipe/analysis/visitAnalysis.py:565]: Yes.  And I hadn't noticed that there's an `ExposureFitsReader.readExposureInfo` that would help, too.

**Pull request merged**